test/index.js

```diff
@@ -105,10 +109,128 @@ describe('KeyringController', function () {
       await keyringController.persistAllKeyrings();
 
       const { vault } = keyringController.store.getState();
-      const keyrings = await mockEncryptor.decrypt(password, vault);
+      const keyrings = await mockEncryptor.decrypt(PASSWORD, vault);
       expect(keyrings.indexOf(unsupportedKeyring) > -1).toBe(true);
       expect(keyrings).toHaveLength(2);
     });
+
+    describe('when `cacheEncryptionKey` is enabled', function () {
+      it('should save an up to date encryption salt to the `memStore` when `password` is unset and `encryptionKey` is set', async function () {
+        keyringController.cacheEncryptionKey = true;
+        const vaultEncryptionKey = '🔑';
+        const vaultEncryptionSalt = '🧂';
+        const vault = JSON.stringify({ salt: vaultEncryptionSalt });
+        keyringController.store.updateState({ vault });
+
+        expect(keyringController.memStore.getState().encryptionKey).toBeNull();
+        expect(
+          keyringController.memStore.getState().encryptionSalt,
+        ).toBeUndefined();
+
+        await keyringController.unlockKeyrings(
+          undefined,
+          vaultEncryptionKey,
+          vaultEncryptionSalt,
+        );
+
+        expect(keyringController.memStore.getState().encryptionKey).toBe(
+          vaultEncryptionKey,
+        );
+        expect(keyringController.memStore.getState().encryptionSalt).toBe(
+          vaultEncryptionSalt,
+        );
+
+        const response = await keyringController.persistAllKeyrings();
+
+        expect(response).toBe(true);
+        expect(keyringController.memStore.getState().encryptionKey).toBe(
+          MOCK_HARDCODED_KEY,
```

Comment:
It's surprising to me that the key would change as a result of `persistAllKeyrings` being called. Having trouble following what's going on here.

Reply:
The password is set in the `beforeEach`. The better fix for this would be to move the controller initialization to a setup function, and call it explicitly in each test so that we can customize it when we need to.

But as a simpler workaround, you could unset the password at the beginning of the test, as we're already doing in the "triggers an error when trying to persist without password or encryption key" test

Reply:
That makes sense, Mark. Thanks a lot for catching that